### ISSUE TITLE
Suspense - onExited sometimes can fail to fire if component suspends at unlucky time

### ISSUE THREAD

**Issue by joshribakoff-sm** [reactjs/react-transition-group#817]
> What is the current behavior?

Around 0.1% of the time [in React 18, not sure if related] `onExited` doesn't fire

> What is the expected behavior?

`onExit` should always fire

> Could you provide a [CodeSandbox](https://codesandbox.io/) demo reproducing the bug?

Unfortunately I have only been able to reproduce this in a larger app using Material UI.

We have intermittent bugs where the MUI `Menu` randomly leaves a `Backdrop` covering our whole page with 0% opacity. 

Through debugging what I originally thought was an MUI issue in the React devtools, I can see MUI fails to hit this condition https://github.com/mui/material-ui/blob/master/packages/mui-base/src/PopperUnstyled/PopperUnstyled.js#L202-L204 

The only possible explanation (that I can think of) is that `react-transition-group` fails to inform the consuming MUI component that the animation is done. The result is that my backdrop has 0% opacity, but MUI never unmounts it, so it steals clicks from the whole page and appears our app is "frozen".

I noticed here is where you set the `EXITING`, https://github.com/reactjs/react-transition-group/blob/master/src/Transition.js#L272-L280 I'm not sure I follow this callback logic, but here you conditionally bail out of the callback https://github.com/reactjs/react-transition-group/blob/master/src/Transition.js#L302-L307 it must be related, I am speculating. Here is where I think it's canceling the callback https://github.com/reactjs/react-transition-group/blob/master/src/Transition.js#L190-L192 and the reason that line runs is that when I select a menu item I update state in my app that triggers the parent to suspend. If it happens to suspend in between EXITING and EXITED, it cancels the callback. Later on when React 18 restores the tree, it restores it to the EXITING state but there is no longer any callback pending to move it towards the terminal state of EXITED, therefore I'm left with this artifact in my UI of a div with 0% opacity preventing clicks.

![Screen Shot 2022-04-13 at 11 55 48 AM](https://user-images.githubusercontent.com/97546767/163251231-bcff3335-1da6-492e-8391-0aec2c9b4c01.png)

**Comment by joshribakoff** [comment]
This still happens after turning off strict mode fyi

**Comment by OllyHodgson** [comment]
I ran into a very similar issue, where in certain scenarios the transition never reaches the `exited` state.

In my case it happens when navigating to a new route where suspense and lazy got involved (i.e. the route components are lazy-loaded). For me the workaround/fix was to pass a `key` attribute to the `Transition` component, e.g.

```jsx
// Get a key from Reach Router
const { key } = useLocation();

return (
  <Transition key={key} 
  {/* etc... */}
```

Unfortunately I wasn't able to produce a reduced test case - it always worked fine in a simpler environment 🤯

**Comment by joshribakoff-sm** [comment]
> Unfortunately I wasn't able to produce a reduced test case

Yeah same. I'm hoping the maintainers can assume this is a valid bug and the burden can be on proving the logic correct, the logic is hard to follow and "looks like" it probably has race conditions.

The way I solved this was to get rid of the MUI menus, and write my own. I wrote my own transitions, too. In my transition, I use a little known React trick where it's actually valid to call `setState` directly in render, so I have logic like this:

```ts
  if (state === 'closed' && in) {
    setState('opening');
  } else if (state === 'opened' && !in) {
    setState('closing');
  }
```

I'm also using `react-spring` and I use the [onRest](https://react-spring.io/common/props#events) callback prop to set the `opened` and `closed` state when the animation ends, depending on the state of the animation. I'm being facetiously snarky here, but in like 6 lines of code I eliminated all the nested callbacks and race conditions that `react-transition` has (in reality I took on a lot of maintenance burden like the CSS for the transitions, etc).  My point is, I don't think it would be hard to rethink the state machine and avoid the callback hell

**Comment by circlingthesun** [comment]
I've tried to created a minimal reproduction but failed. Here is a starting point for anyone else who'd like to try: https://codesandbox.io/s/stoic-gates-qvu5lh?file=/src/App.js

**Comment by haxxxton** [comment]
this might help with y'all above, as i was encountering the same issue.

My application is such that i have a `Context` around my routes that would store an array of alerts.  I then had an `AlertRoot` component that would render `Alert`s based upon the data coming from the `Context` _within_ my base route. When an alert is dismissed it would tell the `Context` it's no longer the "top" alert, changing the `Transition` `in` prop, with an `onExited` function that would tell the `Context` to remove the alert from its array. Like so:

```js
// ./index.jsx
<AlertProvider>
  <RouterProvider router={myRouter} />
</AlertProvider>
```

```js
// ./components/LayoutWrapper.jsx
<div>
  {children}
  <AlertRoot />
</div>
```

```js
// ./routes/*.jsx
// ie. any route component
<LayoutWrapper>
  {/* route specific content */}
</LayoutWrapper>
```

```js
// ./components/AlertRoot
<div>
  {alerts.map((alert) => (
    <Alert
      key={alert.key}
      in={alert.key === topAlert} // <-- when false, begins Transition exit
      onExited={() => deleteAlert(alert.key)} // <-- callback when Transition onExited
    />
  ))}
  {alerts.length > 0 && <Backdrop />}
</div>
```

This worked perfectly, until introduced an alert that would change the route (eg. "our privacy policy has changed, would you like to view it? [Yes] [No]", where "yes" would take the user to `/privacy`). What was happening, was that `AlertRoot` would unmount (which in turn would unmount the `Transition`), _before_ the transition could complete, not firing the `onExited` callback, and leaving my `Backdrop` component rendered.

I managed to solve this by removing `AlertRoot` from `LayoutWrapper`, and moving it to `./index.jsx`. This prevents `AlertRoot` from being unmounted when the route changes, allowing `onExited` to fire, as the `Transition` gets to complete properly

```diff
// ./index.jsx
<AlertProvider>
  <RouterProvider router={myRouter} />
+  <AlertRoot />
</AlertProvider>

// ./components/LayoutWrapper.jsx
<div>
  {children}
-  <AlertRoot />
</div>
```

Unsure if this will help anyone, as your mileage may vary, but thought it might be worth mentioning

**Comment by joshribakoff-sm** [comment]
> What was happening, was that AlertRoot would unmount (which in turn would unmount the Transition), before the transition could complete, not firing the onExited callback, and leaving my Backdrop component rendered.

Interesting theory, but if the transition was unmounted why was the backdrop also not unmounted? Transition or not, portal or not, if the route unmounts all of it's children should also unmount. I think because this library uses too many "escape hatches" (or techniques that may have been only necessary in older versions of React)

**Comment by haxxxton** [comment]
@joshribakoff-sm , in the example code i showed `AlertRoot` is rendered "outside" the route rendering location (after being moved to the `./index.jsx` file).

The render logic i had within `AlertRoot` says "render a backdrop if there are any alerts (regardless of if they are "in")", and the `onExited` callback is what actually deletes the alert from my list of alerts.

In my initial code, this was leaving alerts in my array undeleted, with no alert being "active" (ie. `topAlert`)

I can't speak to the escape hatches, in my case this felt like user error on my part. Perhaps a "unmounted before onExited could fire" warning while in development mode may help to "catch" these sorts of issues?

**Comment by mihanizm56** [comment]
Get the following error with react 18. Any updates?

**Comment by joacub** [comment]
any updates in this ??

**Comment by tomas-c** [comment]
I'm also getting this issue. 

Spent some time investigating and discovered that for some reason when the transition ends, `this.setState` on this line [here](https://github.com/reactjs/react-transition-group/blob/2989b5b87b4b4d1001f21c8efa503049ffb4fe8d/src/Transition.js#L313) may not actually update the state or trigger the callback provided as the second argument. 

The only case where I would expect React to ignore `this.setState` is when the component is being unmounted. But in my case componentWillUnmount is not called so must be something else.

**Comment by AmirABody** [comment]
any updates?!

**Comment by AmirABody** [comment]
I managed to fix the issue temporarily with a forceUpdate (inside setTimeout) in onExiting callback of my Fade Transition.

Sounds like the problem didn't occur anymore, but during code tracing I was completely puzzled why the problem is happening. It's not also related to bail out mechanism in this [line](https://github.com/reactjs/react-transition-group/blob/master/src/Transition.js#L302-L307).

**Comment by joshribakoff** [comment]
It might be a bug in react but unlikely they will fix the legacy api. Seems like this project is dead otherwise it would not be using the legacy API instead of hooks.
